build/common/bitbake_util.py

```diff
@@ -0,0 +1,158 @@
+
+        Args:
+        * obj (PathLike): the path that doesn't exist or invalid
+
+        """
+        super().__init__("{} doesn't exist or invalid.".format(obj))
+
+
+class Bitbake(object):
+    """Class for creating the Bitbake environment."""
+
+    def __init__(
+        self,
+        builddir,
+        outputdir,
+        machine,
+        distro,
+        init_env_file="setup-environment",
+        env_variables=None,
+    ):
+        """
+        Initialize an object of this class.
+
+        Args:
+        * builddir (str): base Bitbake build directory
+        * outputdir (str): directory where to save artifacts
+        * machine (str): Bitbake machine
+        * distro (str): Bitbake distribution
+        * init_env_file (str): initialization environment file
+        * env_variables (dict): the dictionary specified the environment
+                                variables to pass to the Bitbake init file
+
+        """
+        self.builddir = Path(builddir)
+        self.outputdir = Path(outputdir)
+        self.machine = machine
+        self.distro = distro
+        self.init_env_file = init_env_file
+        self.env_variables = env_variables
+        self.shell = bash()
+        self._check_environment()
+
+    def setup_environment(self):
+        """Set up the Bitbake environment."""
+        self.run_command("cd {}".format(self.builddir))
+        env_vars = self._build_env_variables_string()
+        command = "{} MACHINE={} . {} build-{}".format(
+            env_vars, self.machine, self.init_env_file, self.distro
+        )
+        self.run_command(command)
+
+    def run_command(self, command, timeout=None, stdout=True):
+        """
+        Run a command in the Bitbake environment.
+
+        Mandatory args:
+        * command (str): the command to run in the environment
+
+        Optional args:
+        * timeout (int): how long to wait for the next prompt. None means to
+                         wait indefinitely
+        * stdout (bool): flag to print the output to stdout
+
+        Return:
+        * output (str): output of the specified command
+
+        """
+        output = self.shell.run_command(command, timeout=timeout, async_=False)
+        if stdout:
+            print(output)
+        return output
+
+    def run_commands(self, commands, timeout=None, stdout=True):
+        """
+        Run a list of commands in the Bitbake environment.
+
+        Mandatory args:
+        * commands (list): the commands to run in the environment
+
+        Optional args:
+        * timeout (int): how long to wait for the next prompt. None means to
+                         wait indefinitely
+        * stdout (bool): flag to print the output to stdout
+
+        If a command fails its execution, it will continue executing the
+        remaining commands.
+
+        """
+        for command in commands:
+            self.run_command(command, timeout=timeout, stdout=stdout)
+
+    def save_artifacts(self, *files):
```

Comment:
Should this `save_artifacts` method be in this `Bitbake` class? This isn't something BitBake actually does right? This is something our build tools do? I'd recommend making this a free function, possibly in the "file utils" thing, or in its own module.

Reply:
I kind of agree. The save_artifact method was just a convenient method to copy files/directories to an output directory. At the moment we are not even using it. I could remove it altogether and then add it back when we need it. What do you think?

Reply:
Addressed: d42afb6abe63d39574348df7ee963cc370f383ab

Reply:
I think it's useful to have the `save_artifact` function, but I don't really think it belongs in this class. I guess we can just reintroduce it when it's needed. Resolving.